src/base/listener.ts

```diff
@@ -50,14 +50,52 @@ async function attemptClose(maConn: MultiaddrConnection) {
 
 type NodeEntry = {
   latency: number
-  peerId: string
+  peerId?: string
   multiAddr: Multiaddr
 }
 
 function latencyCompare(a: NodeEntry, b: NodeEntry) {
   return a.latency - b.latency
 }
 
+function removeNodeFromList(
+  nodeList: NodeEntry[],
+  maPeerId: string | null,
+  maTuples: ReturnType<Multiaddr['tuples']>
+): NodeEntry[] {
+  const result = []
+
+  for (const entry of nodeList) {
+    const tuples = entry.multiAddr.tuples()
+
+    // Check if same peerId -> duplicate
+    if (maPeerId != null && maPeerId === entry.peerId) {
+      continue
+    }
+
+    // Check if same address:port
+    if (
+      u8aEquals(maTuples[0][1] as Uint8Array, tuples[0][1] as Uint8Array) &&
+      u8aEquals(maTuples[1][1] as Uint8Array, tuples[1][1] as Uint8Array)
+    ) {
+      continue
+    }
+
+    result.push(entry)
+  }
+
+  return result
+}
+
+function isUsableRelay(tuples: ReturnType<Multiaddr['tuples']>, maPeerId: string | null, self: PeerId) {
```

Comment:
Instead of passing esoterical params, can we just pass a `Multiaddr` of a relay?

Reply:
done